Add initial simulation code

Chippiewill opened the pull request:
Add simulation skeleton, GUI, and dummy servers, to show an example and
provide general architecture

uptonfunk committed:
Add initial simulation code

Add simulation skeleton, GUI, and dummy servers, to show an example and
provide general architecture

uptonfunk committed:
Add initial simulation code

Add simulation skeleton, GUI, and dummy servers, to show an example and
provide general architecture

uptonfunk committed:
enable cars to listen for deallocation while sending location heartbeat

Revised some long lines, splitting them up.
Added functionality for the simulation to be stopped for testing

HarveyHunt committed:
serialise: Rename REST models and remove error field from response

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

HarveyHunt committed:
rest: Initial work on REST API

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

HarveyHunt committed:
rest: Add REST server tests

Signed-off-by: Harvey Hunt <harveyhuntnexus@gmail.com>

uptonfunk committed:
remove obsolete test

uptonfunk committed:
remove obsolete test

Commit:
Merge pull request #11 from HarveyHunt/rest-api-pr

Initial REST API Implementation

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+
+        count = 0
+        name = 0
+        while count < no_spaces:
```
This appears to have no way of _stopping_ a simulation. 
It would be good to move the actual loop into a coroutine with a yield point on each loop.

For testing it would be nice to be able to say something like:

    sim = SimManager(*state)
    asyncio.ensure_future(sim.run())
    await asyncio.sleep(3)
    await sim.stop()
    assert some_test_on_the_sim_state(sim)
   
Then we could run simulations, stop them and extract data from it afterwards.

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+        return self.aDestX, self.aDestY
+
+
+class ParkingLot:
```
There's a lot of duplication here with `parking.shared.rest_models.ParkingLot`

This could just subclass that and add the simulated data, and the allocation methods (or better yet, the `SimManager` should perform the allocation and just update the stored parking lot object).

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+    def get_location(self):
+        return self.lat, self.long
+
+    def get_capacity(self):
```
`get_` methods aren't needed in python, you can just replace a `lot.get_capacity()` with `lot.capacity` already here.

bensimner commented on tests/test_Parking.py:
```diff
@@ -0,0 +1,51 @@
+#                                            })
+
+
+def test_AddZeroSpaces():
```
If you re-use `parking.shared.rest_models.ParkingLot` then these tests would be unnecessary. :)

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+            else:
+                n = self.random_lot.randint(min_spaces_per_lot, max_al)
+            price = round(self.random_lot.uniform(0, 10), 2)
+            self.space_tasks.append(asyncio.ensure_future(space_routine(0,
```
This formatting is kinda ugly, how about something like:

    coro = space_routine(0, px, py, n, str(name), price, n, self)
    self.space_tasks.append(asyncio.ensure_future(coro))

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+
+class SimManager:
+    def __init__(self, no_spaces, min_spaces_per_lot, max_spaces_per_lot, no_cars,
+                 x, y, parking_lot_seed, car_seed, max_time):
```
What are `x` and `y`? Are they the width/height of the grid?

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+        self.waypoints = []
+        self.waypoints.append(Waypoint(time.time(), self.lat, self.long))
+
+    def get_location(self):
```
As with `ParkingLot` you don't need `get_` methods, you can just make `Car` store a `parking.shared.location.Location` object and store it in `self.location`

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+            name += 1
+
+        for i in range(self.no_cars):
+            self.car_tasks.append(asyncio.ensure_future(car_routine(round(self.random_car.uniform(0, self.max_time), 1),
```
As before this formatting can be split into two lines to avoid adding weird linebreaks that are hard to read.

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+    car = Car(startx, starty)
+    manager.cars.append(car)
+
+    async with websockets.connect('ws://localhost:8765') as websocket:
```
We should probably create a client-like object that `car_routine` and `space_routine` can use, then they could look something like:

    async def car_routine(manager, x, y):
        # create car
        car = Car(x, y)
        manager.cars.append(car)
        
        # set initial state
        car.set_state(Car.ROAMING)
    
        async with client() as cl:
            cl.send(wsmodels.ParkingRequestMessage(car.destination))
            
            while True:
                # see if we recieved a message from the server
                # and update the Car's state if we did.
                await msg = cl.recv(blocking=False)
                if isinstance(msg, wsmodels.ParkingAllocationMessage):
                    car.set_state(Car.ALLOCATED_PARKING)
                    car.allocated_lot = lot
                elif ...:
                    ...

bensimner commented on simulation/simulation.py:
```diff
@@ -0,0 +1,295 @@
+    def fill_space(self) -> bool:
+        if self.available > 0:
+            msgbody = serialize_model(restmodels.SpaceAvailableMessage(self.available - 1))
+            request = httpclient.HTTPRequest(resturl + "/:" + str(self.id) + "/available", body=msgbody, method='POST')
```
I feel like the methods of a parking lot object shouldn't try do http requests.

The car/space routines should do that and then update the parking lot objects on the SimManager.

bensimner reviewed (changes requested):
There's some good stuff here. With a few requests:

- There's not really any tests for the code here.
- Moving http logic into the car/space_routines will make integration easier
- Car's will eventually get more state, i.e. roaming/allocated/whatever and they will need to handle that.

I suggest making the car/space_routine do all the http interfacing, and the Car/ParkingLot objects have their own step methods that get called by the SimManager in a loop.

uptonfunk committed:
refactor networking, simulation manager and code formatting

uptonfunk committed:
add a helper class to separate networking from parking, update tests

Chippiewill committed:
Refactor parking lot REST API client, convert more REST API tests

This tidies up the previous commit a little by converting more
REST API tests. It also renames the `parking_rest_helper` module to
the more general 'clients' (shorter, and allows for other clients
to be stored in the same module which is more pythonic) lastly
it restricts some of the checking done by the client itself to
the minimum needed.

uptonfunk committed:
add a client class to separate networking from cars

uptonfunk committed:
add tests for ws client

uptonfunk committed:
added websockets to requirements

Chippiewill committed:
Simplify attribute type checking

- Removes some redundant checks (e.g. type checking after conversion)
- Moves some checks to conversions (e.g. checking for float -> convert to float)
- Uses `Attribute.type` for explicit type checks via `utils.enforce_type`
-

manfredlift committed:
Merge pull request #16 from Chippiewill/master

Simplify attribute type checking

Commit:
Remove pointless testing (#17)

uptonfunk committed:
add explicit send and receive for each ws message type

uptonfunk committed:
Update tests

uptonfunk committed:
Update to use shared classes, including web API client classes, minor improvements

uptonfunk committed:
removed obsolete tests

uptonfunk committed:
Merge remote-tracking branch 'origin/simulation' into simulation

# Conflicts:
#	requirements.txt
#	simulation/simulation.py
#	simulation/testserver.py
#	simulation/testspaceserver.py

uptonfunk committed:
enable cars to listen for deallocation while sending location heartbeat

uptonfunk committed:
minor fix in error handling and test server

MattDevy committed:
Revised some long lines, splitting them up.
Added functionality for the simulation to be stopped for testing

uptonfunk committed:
removed obsolete test

uptonfunk committed:
Merge remote-tracking branch 'origin/simulation' into simulation
